crates/trie/trie/src/node_iter.rs

```diff
@@ -84,28 +93,43 @@ impl<C, H: HashedCursor> TrieNodeIter<C, H> {
 
     /// Seeks the hashed cursor to the given key.
     ///
+    /// If the key is the same as the last seeked key, the result of the last seek is returned.
+    ///
     /// If `metrics` feature is enabled, also updates the metrics.
-    fn hashed_cursor_seek(&mut self, key: B256) -> Result<Option<(B256, H::Value)>, DatabaseError> {
+    fn seek_hashed_entry(&mut self, key: B256) -> Result<Option<(B256, H::Value)>, DatabaseError> {
+        if let Some(entry) = self
+            .last_seeked_hashed_entry
+            .as_ref()
+            .filter(|entry| entry.seeked_key == key)
+            .map(|entry| entry.result)
+        {
+            #[cfg(feature = "metrics")]
+            self.metrics.inc_leaf_nodes_same_seeked();
+            return Ok(entry);
+        }
+
+        let result = self.hashed_cursor.seek(key)?;
+        self.last_seeked_hashed_entry = Some(SeekedHashedEntry { seeked_key: key, result });
+
         #[cfg(feature = "metrics")]
         {
             self.metrics.inc_leaf_nodes_seeked();
 
-            if Some(key) == self.previously_seeked_key {
-                self.metrics.inc_leaf_nodes_same_seeked();
-            }
-            self.previously_seeked_key = Some(key);
-
-            if Some(key) == self.previously_advanced_to_key {
+            // If either the seek key or the result key is the same as the key we previously
+            // advanced to.
+            if Some(key) == self.previously_advanced_to_key ||
+                result.as_ref().map(|(k, _)| *k) == self.previously_advanced_to_key
+            {
```

Comment:
not sure I understand this addition, how do we expect this to change the metrics w/ the new changes?

Reply:
basically, this metric now reports the number of times we re-used the previous seeked entry. We're not tracking the original meaning of this metric anymore, because it's just at zero at all times.

Reply:
reverted